DOCS-1195: Review modular resources (#1992)
viamrobotics/docs: docs/modular-resources/create/_index.md

  ]
description: "Use the Viam module system to implement modular resources that can be included in any Viam-powered smart machine."
no_list: true
aliases:

>   - "/extend/modular-resources/create/"

Reviewer: we do need this

You can extend the capabilities of Viam, by creating a custom {{< glossary_tooltip term_id="module" text="module" >}} that contains one or more {{< glossary_tooltip term_id="modular-resource" text="modular resources" >}}.
You can integrate modules into any Viam-powered smart machine.

> One common scenario for using {{< glossary_tooltip term_id="modular-resource" text="modular resources" >}} is to create a new [model](/extend/modular-resources/key-concepts/#models) that implements an existing Viam [API](/program/apis/).

Reviewer: ```suggestion
One common scenario for using {{< glossary_tooltip term_id="modular-resource" text="modular resources" >}} is to create a new [model](/modular-resources/key-concepts/#models) that implements an existing Viam [API](/program/apis/).
```

> Once you have created your modular resource, you can use the [Viam CLI](/manage/cli/) to [upload your modular resource](/extend/modular-resources/upload/) to the [Viam registry](https://app.viam.com/registry) to share it with other Viam users or just to other users in your organization.
> For added convenience, you can configure [automated uploads for new module versions](/extend/modular-resources/upload/#update-an-existing-module-using-a-github-action) through a continuous integration (CI) workflow, using a GitHub Action.

Reviewer: ```suggestion
Once you have created your modular resource, you can use the [Viam CLI](/manage/cli/) to [upload your modular resource](/modular-resources/upload/) to the [Viam registry](https://app.viam.com/registry) to share it with other Viam users or just to other users in your organization.
For added convenience, you can configure [automated uploads for new module versions](/modular-resources/upload/#update-an-existing-module-using-a-github-action) through a continuous integration (CI) workflow, using a GitHub Action.
```

> You can also add your module to your robot as a [local module](/extend/modular-resources/configure/#local-modules), without uploading it to the Viam registry.

Reviewer: ```suggestion
You can also add your module to your robot as a [local module](/modular-resources/configure/#local-modules), without uploading it to the Viam registry.
```

## Prepare to code a new resource model

> Before you start coding a new resource model, follow these steps to find the appropriate source code:

Reviewer: ```suggestion
Before you start coding a new resource model, follow these steps to find the appropriate source code as a reference for the methods you need to implement:
```

1. **Understand your model's client interface:**

   To create a new resource model, you need to implement your model's "**client** interface" in a file called `my_modular_resource.go` or `my_modular_resource.py`.

>    This interface defines how your model's server will responds to API requests.

Reviewer: ```suggestion
   This interface defines how your model's server will respond to API requests.
```

1. **Prepare to Import Your Custom Model and Subtype's API:**

   To prepare to import your custom model and chosen resource subtype's API into your main program and register them with your preferred SDK:

   - Find the subtype API as defined in the relevant `<resource-name>/<resource-name>.go` file in the RDK on Viam's GitHub.
     - For example, the base subtype is defined in [<file>rdk/components/base/base.go</file>](https://github.com/viamrobotics/rdk/blob/main/components/base/base.go#L37).
   - Reference <file>main.go</file> or <file>main.py</file> when editing this second file.

## Create a custom module

A custom module can implement one or more [models](/extend/modular-resources/key-concepts/#models).
To create a custom module, follow the steps below:

1. [Code a new resource model](#code-a-new-resource-model) server. It must have all the methods that the Viam RDK requires, and should match the built-in API client {{< glossary_tooltip term_id="subtype" text="subtype" >}} like ['rdk:component:base'](https://pkg.go.dev/go.viam.com/rdk@v0.10.0/components/base#pkg-functions).

>    Save this in a file within your module’s <file>my_modular_resource.go</file> or <file>my_modular_resource.py</file> file.

Reviewer: I think the model naming has gone missing
```suggestion
   Save this in a file within your module’s <file>my_modular_resource.go</file> or <file>my_modular_resource.py</file> file.
   Name your model according to the namespace of the built-in API you are implementing.
   For more information see [Naming your model](/modular-resources/key-concepts/#naming-your-model-namespacerepo-namename).
```

2. [Code a main program](#code-a-main-entry-point-program), <file>main.go</file> or <file>main.py</file>, that starts the module after adding your desired resources from the registry.
   Import your custom model and API into the main program and register them with your chosen SDK.
   This main program is the "entry point" to your module.

3. [Compile or package](#compile-the-module-into-an-executable) the module into a single executable that can receive a socket argument from Viam, open the socket, and start the module at the entry point.

### Code a new resource model

> The following example module registers a modular resource implementing Viam's built-in [Base API](/components/base/#api) [(rdk:service:base)](/extend/modular-resources/key-concepts/#models) as a new model, `"mybase"`, using the model family `acme:demo:mybase`.
> 
> - For more information, see [Naming your model](/extend/modular-resources/key-concepts/#naming-your-model).

Reviewer: ```suggestion
The following example module registers a modular resource implementing Viam's built-in [Base API](/components/base/#api) [(rdk:service:base)](/modular-resources/key-concepts/#models) as a new model, `"mybase"`, using the model family `acme:demo:mybase`.

- For more information, see [Naming your model](/modular-resources/key-concepts/#naming-your-model).
```

from viam.resource.registry import Registry, ResourceCreatorRegistration
from viam.resource.types import Model, ModelFamily
from viam.utils import ValueTypes
from viam.logging import getLogger

> LOGGER = getLogger(__name__)

Reviewer: This should also stay

Reviewer: This is still gone and shouldn't be

from viam.resource.registry import Registry, ResourceCreatorRegistration
from viam.resource.types import Model, ModelFamily

> from viam.utils import ValueTypes

Reviewer: ```suggestion
from viam.utils import ValueTypes
from viam.logging import getLogger

LOGGER = getLogger(__name__)
```

You must define all functions belonging to a built-in resource subtype's API if defining a new model.
Otherwise, the class won’t instantiate.

> - If you are using the Python SDK, raise an `NotImplementedError()` in the body of functions you do not want to implement or put `pass`.

Reviewer: ```suggestion
- If you are using the Python SDK, raise an`NotImplementedError()` in the body of functions you do not want to implement or put `pass`.
```

Reviewer: ditto you didn't address this

- If you are using the Go SDK, return `errUnimplemented`.
- Additionally, return any values designated in the function's return signature, typed correctly.

{{% /alert %}}

### Compile the module into an executable

> To [add a module](/extend/modular-resources/configure/) to the configuration of your robot, you need to have an [executable](https://en.wikipedia.org/wiki/Executable) that runs your module when executed, can take a local socket as a command line argument, and cleanly exits when sent a termination signal.

Reviewer: ```suggestion
To [add a module](/modular-resources/configure/) to the configuration of your robot, you need to have an [executable](https://en.wikipedia.org/wiki/Executable) that runs your module when executed, can take a local socket as a command line argument, and cleanly exits when sent a termination signal.
```

You need to ensure any dependencies for your module (including the Python or Go [Viam SDK](/program/)) are installed as well.
Your executable will be run by `viam-server` as root, so dependencies need to be available to the root user.

> ### Configure logging

Reviewer: I think we probably want to keep this ;)

> Once you have created your custom resource, you can use the [Viam CLI](/manage/cli/) to [upload your custom resource](/extend/modular-resources/upload/) to the Viam registry, to share it with other Viam users or just to other users in your organization.

Reviewer: ```suggestion
Once you have created your custom resource, you can use the [Viam CLI](/manage/cli/) to [upload your custom resource](/modular-resources/upload/) to the Viam registry, to share it with other Viam users or just to other users in your organization.
```

> Alternatively, you can [add your module locally](/extend/modular-resources/configure/) to your robot without uploading to the Viam registry.

Reviewer: ```suggestion
Alternatively, you can add your module to your robot as a [local module](/modular-resources/configure/#local-modules) without uploading it to the Viam registry.
```
